### PR TITLE
fix opening a floating view on the NOOP output

### LINKED ISSUE #3954: segfault in container_init_floating
* Sway Version: 1.0

* Debug Log: https://pub.yori.cc/sway-crash10.log

* Configuration File: https://pub.yori.cc/i3_config

Finally, explain the steps you took in plain English to reproduce the problem below.

1. disconnect from docking station

* Stack Trace, if sway crashes:
```
#0  0x000000000044819f in container_init_floating ()
#1  0x0000000000449e32 in container_set_floating ()
#2  0x000000000044dd61 in view_map ()
#3  0x00000000004235e3 in handle_map ()
#4  0x00007f96a300c9fc in wlr_signal_emit_safe ()
   from /nix/store/qy5an0g3krz886f5qvw64b03br1nlmd9-wlroots-408eca7dfa12eda0f1b0ec6050e99ee2e6a8f2b4/lib/libwlroots.so.2
#5  0x00007f96a2ff2049 in handle_xdg_surface_commit ()
   from /nix/store/qy5an0g3krz886f5qvw64b03br1nlmd9-wlroots-408eca7dfa12eda0f1b0ec6050e99ee2e6a8f2b4/lib/libwlroots.so.2
#6  0x00007f96a3006dd2 in surface_commit_pending ()
   from /nix/store/qy5an0g3krz886f5qvw64b03br1nlmd9-wlroots-408eca7dfa12eda0f1b0ec6050e99ee2e6a8f2b4/lib/libwlroots.so.2
#7  0x00007f96a3007088 in surface_commit () from /nix/store/qy5an0g3krz886f5qvw64b03br1nlmd9-wlroots-408eca7dfa12eda0f1b0ec6050e99ee2e6a8f2b4/lib/libwlroots.so.2
#8  0x00007f96a27cd05e in ffi_call_unix64 () from /nix/store/kbbiky87bsdndh63m7919sxkf3gsh5zq-libffi-3.2.1/lib/../lib64/libffi.so.6
#9  0x00007f96a27cc023 in ffi_call () from /nix/store/kbbiky87bsdndh63m7919sxkf3gsh5zq-libffi-3.2.1/lib/../lib64/libffi.so.6
#10 0x00007f96a3050b94 in wl_closure_invoke () from /nix/store/9n45jlbvwi49hlmm0kqlgllhnznq1x04-wayland-1.16.0/lib/libwayland-server.so.0
#11 0x00007f96a304d61f in wl_client_connection_data () from /nix/store/9n45jlbvwi49hlmm0kqlgllhnznq1x04-wayland-1.16.0/lib/libwayland-server.so.0
#12 0x00007f96a304ebe2 in wl_event_loop_dispatch () from /nix/store/9n45jlbvwi49hlmm0kqlgllhnznq1x04-wayland-1.16.0/lib/libwayland-server.so.0
#13 0x00007f96a304d7fa in wl_display_run () from /nix/store/9n45jlbvwi49hlmm0kqlgllhnznq1x04-wayland-1.16.0/lib/libwayland-server.so.0
#14 0x000000000040e509 in main ()
```
Release build, built without debug.

### PR DESCRIPTION
**Pull request by RedSoxFan**
Fixes #3954

Since the NOOP output has no size, the minimum floating size is greater
than the workspace size for the NOOP output. In this case, the floater
gets centered in the output instead of the workspace. However, the
NOOP output is not part of the output layout and thus has a NULL box.
Attempting to access the properties of this box was causing a segfault.

This fixes the issue by just setting the floater's box to all zeroes
when mapping on the NOOP output. When the workspace gets moved from the
NOOP output to a new output, any floater whose width or height is zero
or has an x or y location outside of the output, gets passed to
`container_init_floating` again. This will then set the appropriate
size and centering. For any floater that has a valid size and location,
they are preserved.

### CODE REVIEW

File: sway/tree/output.c [hunk 1]
```diff
@@ -60,6 +60,13 @@ static void restore_workspaces(struct sway_output *output) {
 		struct sway_workspace *ws = root->noop_output->workspaces->items[0];
 		workspace_detach(ws);
 		output_add_workspace(output, ws);
+		for (int i = 0; i < ws->floating->length; i++) {
```
Comment (emersion) [inline sway/tree/output.c]: Is it possible to add a comment here to explain in which situation we need to check for `width == 0 || height == 0 || …`?

Comment (RedSoxFan) [reply sway/tree/output.c]: Done

File: sway/tree/output.c [hunk 2]
```diff
@@ -60,6 +60,21 @@ static void restore_workspaces(struct sway_output *output) {
...
+		// If the floater was made floating while on the NOOP output, its width
+		// and height will be zero and it should be reinitialized as a floating
+		// container to get the appropriate size and location. Additionally, if
+		// the floater's top-left corner is outside of the workspace, do the
+		// same as the output layout has likely changed and the maximum size
+		// may be different
+		for (int i = 0; i < ws->floating->length; i++) {
+			struct sway_container *floater = ws->floating->items[i];
+			if (floater->width == 0 || floater->height == 0 ||
+					floater->x < ws->x || floater->y < ws->y) {
```
Comment (emersion) [inline sway/tree/output.c]: I don't think this check is right/enough.

Shouldn't we also check for the right/bottom borders? Shouldn't users be able to drag the top-left corner past the workspace bounds?

Comment (RedSoxFan) [reply sway/tree/output.c]: I think it should be good now. I changed it to use the output bounds instead of the workspace bounds and added a check for the bottom/right borders

Comment (emersion) [reply sway/tree/output.c]: One question remains: do we want to allow this configuration?

```
floating
+-------------------------+
|                         |
|                         |
|                         |
|         +------------------------------------------+
|         |               |                          |
|         |               |                          |
|         |               |                          |
|         |               |                          |
+-------------------------+                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          |                                          |
          +------------------------------------------+
          workspace
```

Comment (RedSoxFan) [reply sway/tree/output.c]: I changed it to check for a width/height of zero, a width/height larger than the output, or being completely outside of the output. That should allow for the above to be retained.

**Comment by emersion** [comment]
Thanks!
